docs/metrics-data-model.md

```diff
@@ -0,0 +1,108 @@
+# k6 Time Series
+
+We want to introduce the time series concept to k6 for getting the benefit of efficiently identifying the combination between a metric and the relative tags.
+
+A summary of the benefits it brings across the platform:
+
+* It allows executing faster lookups and comparisons by a single value, the ID (hash) of the series.
+* It reduces the storage by allocating just one time a possible combination (the time series) and referencing it in all other places. 
+* It allows aggregating data by time series.
+* It enables easier detection of high cardinality issues just by counting the entire unique data set of generated time series.
+* It enables time series based outputs to simplify the integration's logic (e.g. Prometheus).
+
+## Data model
+
+#### Metrics and Series
+
+* Metric (name, type)
+* Time Series (metric, tags)
+* Sample (timestamp, value, time series ref)
+* Tag - (aka [k6/1831](https://github.com/grafana/k6/issues/1831))
+
+```go
+// This is the current k6 Metric polished from the Thresholds dependencies.
+type Metric struct {
+	name     string
+	type     MetricType
+	contains ValueType
+}
+
+type Tag struct {
+    Key, Value string
+}
+
+type TimeSeries struct {
+	id uint64 // hash(metric+tags)
```

Comment:
```suggestion
	id uint32 // sequential id, internal for the 
	hash uint64 // hash(metric+tags)
```

Since we'll have this central thread-safe `Repository` / `Registry`, with locking and everything, we can assign each TimeSeries a unique (in the instance) sequential ID as well, without any overhead :thinking: This might be useful in outputs, since it could potentially allow for faster caches. Instead of a map lookup on every metric, it we could use slices, since these outputs will expect to see metrics with all generated time series anyway. It needs to be benchmarked, since maps are pretty quick as well, but it might help and won't be costly (in either performance or labor) to add :man_shrugging:

Reply:
It could be a very good point of optimization. The access to a slice by index should be at least an order of magnitude faster.

I'm just not convinced about the name. If we would have an ID of the TimeSeries I think we should use the `Hash` because it's consistent across instances and runs, it is also not removable from the design because it is an essential concept. So I would prefer to use another name like `SequenceNum{ber}` / `LocalID` / `Index`.

There is a 3rd option where we don't use ID and we use only Hash and another name (e.g `Sequence`).

Reply:
Hmm :thinking: How about no ID field at all, to avoid any possible confusion? For example: `hash` and `seqNumberInInstance`?